tinygrad/shape/shapetracker.py

```diff
@@ -100,30 +100,54 @@ def merge_views(vm2:View, vm1:View) -> Optional[View]:
+    elif old_dim < new_dim: # reshaping a mask to a larger dimension
+      next_mask = next(r_masks, (0,1)) 
+      # if the current dimension is masked, we cannot merge unless the next mask has a range of 1
+      if (mask[0]!=0 or mask[1]!=old_dim) and next_mask[1]-next_mask[0]!=1: return view.mask, True
+      mask = (next_mask[0]*old_dim+mask[0], (next_mask[1]-1)*old_dim+mask[1])
+      old_dim *= next(r_shape, 1) # we merge this dimension with the next and try again
+    else: # old_dim > new_dim
+      if (mask[0]%new_dim!=0 or mask[1]%new_dim!=0) and mask[0]//new_dim!=(mask[1]-1)//new_dim: # cannot split if the reshape cuts across the mask
+        return view.mask, True
+      new_mask.append((mask[0]%new_dim//stride, (((mask[1]-1)%new_dim+1)-1)//stride+1))
+      stride *= new_dim
+      new_dim *= next(r_new_shape, 1)
+  for mask in (mask, *r_masks): # if the old shape has leading 1s, need to make sure their mask is (0,1)
+    if mask != (0,1): return ((0,0),)*len(new_shape), False
+  return tuple(reversed(new_mask)), False
+
 @functools.lru_cache(maxsize=None)
 def _reshape(view: View, new_shape:Tuple[int, ...]) -> Tuple[View, bool]:
-  shape, mask, strides, offset = view.shape, view.mask, view.strides, view.offset
-  # check if this is adding or removing 1s (only)
-  # NOTE: this is optional, but removes most calls to (expensive!) merge_views (with mask, not optional)
-  if [x for x in shape if x != 1] == [x for x in new_shape if x != 1]:
-    new_strides: List[int] = [y for x,y in zip(shape, strides) if x != 1]
-    new_strides_tuple: Tuple[int, ...] = tuple([0 if x == 1 else new_strides.pop(0) for x in new_shape])
-    new_mask_tuple = None
-    if mask:
-      for x,y in zip(shape, mask):
-        if x == 1 and y != (0, 1):
-          new_mask_tuple = ((0,0),) * len(new_shape)
-          break
-      else:
-        new_mask: List[Tuple[int, int]] = [y for x,y in zip(shape, mask) if x != 1]
-        new_mask_tuple = tuple([(0,1) if x == 1 else new_mask.pop(0) for x in new_shape])
-    return View(new_shape, new_strides_tuple, offset, new_mask_tuple), False
-
-  new_view = View(new_shape, strides_for_shape(new_shape))
-  if view.contiguous: return new_view, False # NOTE: if it's contiguous it can't have an offset
-  if (merged_view := merge_views(view, new_view)) is not None: return merged_view, False
+  if view.contiguous or (view.shape == new_shape): return View(new_shape), False
+  strides, reverse_shape = [], reversed(new_shape)
+  # shape_strides[0][0] == 1 implies view.shape consists only of 1's
+  if view.shape_strides[0][0] == 1: return View(new_shape, offset=view.offset), False
+  for d, s in reversed(view.shape_strides):
+    acc, next_stride = 1, s
+    while acc < d:
+      new_dim = next(reverse_shape)
+      acc *= new_dim
+      strides.append(next_stride)
+      next_stride *= new_dim
+    if acc != d: break
+  else:
+    while len(strides) < len(new_shape): strides.append(0)
```

Comment:
`if len(new_shape) > len(strides) : strides += [0,] * (len(new_shape) - len(strides))`

Reply:
Yeah that looks better, thanks